hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java

```diff
@@ -864,33 +879,19 @@ protected void unblockWhenComplete(@Nonnull final BiConsumer<? super V, ? super
         }
     }
 
-    @Override
-    public InternalCompletableFuture<V> exceptionally(@Nonnull Function<Throwable, ? extends V> fn) {
-        requireNonNull(fn);
+    private void unblockExceptionally(@Nonnull Function<Throwable, ? extends V> fn,
+                                      InternalCompletableFuture<V> future) {
         Object result = resolve(state);
```

Comment:
It is wrong to resolve the `state` at this point and then use its value in a further condition. Instead, we should first check if `AbstractInvocationFuture#isDone`, similarly to all other `CompletionStage` API implementations.